<?xml version="1.0"?>
<launch>

  ############################# Global Param ###################################

  <!--   flight_mode
      *pos mode: pos control with hokuyo + keyboard,
      *vel mode: vel control with px4 + joystick
      the most important param
      controlMode: switch betwen vel and pos
      altitudeControlMode: switch between laser and sonoar
      usefreefall: necessary for snoar mode
      useouteryawest: necessary for lasersalm + pos
  -->

  <arg name="debug" default="false" />
  <arg unless="$(arg debug)" name="launch_prefix" value="" />
  <arg if="$(arg debug)" name="launch_prefix" value="gdb --ex run --args" />


  <arg name="flight_pos_mode" value="false" /> <!-- pos mode -->
  <arg name="flight_vel_mode" value="true" />  <!-- vel mode -->


  ############################# JSK Quadcopter ###################################

  <node pkg="aerial_robot_base" type="aerial_robot_base_node" name="aerial_robot_base_node"  launch-prefix="$(arg launch_prefix)" output="screen" > 

    ############### Basic Param  #####################
    <param name="rx_loop_rate" type="double" value="110"/>
    <param name="tx_loop_rate" type="double" value="40"/>
    <param name="tf_pub_loop_rate" type="double" value="120"/>

    <param name="motor_num" type="int" value="1"/>


    <!-- z axis   -> 0: laser, 1:sonar -->
    <param if="$(arg flight_pos_mode)" name="estimator/altitude_control_mode" type="int" value="0"/>
    <!-- yaw axis -> true for slam; -->
    <param if="$(arg flight_pos_mode)" name="estimator/use_outer_yaw_est" type="bool" value="true"/>
    <!-- x/y axis -> pos control: 0, vel control: 1 -->
    <param if="$(arg flight_pos_mode)" name="navigator/xy_control_mode" type="int" value="0"/>
    <param if="$(arg flight_vel_mode)" name="estimator/altitude_control_mode" type="int" value="1"/>
    <param if="$(arg flight_vel_mode)" name="estimator/use_outer_yaw_est" type="bool" value="false"/>
    <param if="$(arg flight_vel_mode)" name="navigator/xy_control_mode" type="int" value="3"/> <!-- pos local vel -->

    <!--degub -->
    <param name="navigator/gain_tunning_mode" type="int" value="0"/>


    ############### Quadctoper Basic State ######################
    <rosparam file="$(find aerial_robot_base)/config/d_rook/EstimatorConfig.yaml" command="load" />
    ############### PID Control ######################
    <rosparam file="$(find aerial_robot_base)/config/d_rook/PidControlConfig.yaml" command="load" />

    ############### Kalman Filter ####################
    <param name="estimator/kalman_filter_flag" type="bool" value="true"/>
    <param name="kalman_filter/imu_hz" type="double" value="105"/>
    <rosparam file="$(find aerial_robot_base)/config/rook/KduinoKalmanFilterConfig.yaml" command="load" />
    <rosparam file="$(find aerial_robot_base)/config/rook/OpticalFlowKalmanFilterConfig.yaml" command="load" />

    ############### Motion Capture####################
    <param name="estimator/mocap_flag" type="bool" value="true"/>
    <param name="mocap/cog_offset" type="bool" value="false"/>


    <param name="estimator/px4flow_flag" type="bool" value="false"/>

    ############### Keyboard Navigation ##############
    <rosparam file="$(find aerial_robot_base)/config/d_rook/TeleopNavigationConfig.yaml" command="load" />

    ############### remap topic ######################
    ### Optical Flow ##############
    <remap from="optical_flow/opt_flow" to="px4flow/opt_flow" />
    ### Joy Stick #################
    <remap from="/navigator/joy_stick_command" to="/joy" />
    ### Teleop Navigator ##########
    <remap from="/navigator/teleop_command/land" to="/teleop_command/land" />
    <remap from="/navigator/teleop_command/start" to="/teleop_command/start" />
    <remap from="/navigator/teleop_command/halt" to="/teleop_command/halt" />
    <remap from="/navigator/teleop_command/takeoff" to="/teleop_command/takeoff" />
    <remap from="/navigator/teleop_command/roll" to="/teleop_command/roll" />
    <remap from="/navigator/teleop_command/pitch" to="/teleop_command/pitch" />
    <remap from="/navigator/teleop_command/yaw" to="/teleop_command/yaw" />
    <remap from="/navigator/teleop_command/throttle" to="/teleop_command/throttle" />
    <remap from="/navigator/teleop_command/ctrl_mode" to="/teleop_command/ctrl_mode" />

    <remap from="/arming_ack" to="/jsk_stm_serial_comm/arming_ack" />
    <remap from="/aerial_robot_control" to="/jsk_stm_serial_comm/aerial_robot_control" />
    <remap from="/config_cmd" to="/jsk_stm_serial_comm/config_cmd" />

    <remap from="/kduino/rc_cmd" to="/deprected" />
    <remap from="/controller/debug" to="/test/controller/debug" />

  </node>


  ################################# jsk_stm ########################################

<node name="jsk_stm" pkg="jsk_stm" type="jsk_stm_node" respawn="false"  output="screen">
  <param name="serial_port" value="/dev/ttyUSB0" />
</node>

  ################################# Joy Stick ######################################

  <include file="$(find aerial_robot_base)/launch/sensors/joy_stick.launch" />

  ################################# kduino test ########################################


  <node pkg="rosserial_python" type="serial_node.py" name="serial_node" output="screen">
    <param name="baud" value="115200" />
    <param name="port" value="/dev/ttyUSB1" />
  </node>


  <arg name="flight_pos_mode" value="true" /> <!-- pos mode -->
  <arg name="flight_vel_mode" value="false" />  <!-- vel mode -->


  <node pkg="aerial_robot_base" type="aerial_robot_base_node" name="aerial_robot_base_node_test"  launch-prefix="$(arg launch_prefix)" output="screen" > 

    ############### Basic Param  #####################
    <param name="rx_loop_rate" type="double" value="110"/>
    <param name="tx_loop_rate" type="double" value="40"/>
    <param name="tf_pub_loop_rate" type="double" value="120"/>

    <param name="motor_num" type="int" value="1"/>

    <!-- z axis   -> 0: laser, 1:sonar -->
    <param if="$(arg flight_pos_mode)" name="estimator/altitude_control_mode" type="int" value="0"/>
    <!-- yaw axis -> true for slam; false for kduino -->
    <param if="$(arg flight_pos_mode)" name="estimator/use_outer_yaw_est" type="bool" value="true"/>
    <!-- x/y axis -> pos control: 0, vel control: 1 -->
    <param if="$(arg flight_pos_mode)" name="navigator/xy_control_mode" type="int" value="0"/>
    <param if="$(arg flight_vel_mode)" name="estimator/altitude_control_mode" type="int" value="1"/>
    <param if="$(arg flight_vel_mode)" name="estimator/use_outer_yaw_est" type="bool" value="false"/>
    <param if="$(arg flight_vel_mode)" name="navigator/xy_control_mode" type="int" value="3"/> <!-- pos local vel -->

    ############### Quadctoper Basic State ######################
    <rosparam file="$(find aerial_robot_base)/config/d_rook/EstimatorConfig.yaml" command="load" />
    ############### PID Control ######################
    <rosparam file="$(find aerial_robot_base)/config/rook/PidControlConfig.yaml" command="load" />

    ############### _motion Capture####################
    <param name="mocap/cog_offset" type="bool" value="false"/><!-- new for new hydra control -->
    <param name="estimator/mocap_flag" type="bool" value="true"/>
    <param name="estimator/px4flow_flag" type="bool" value="false"/>

    ############### Kalman Filter ####################
    <param name="estimator/kalman_filter_flag" type="bool" value="false"/>

    ############### Keyboard Navigation ##############
    <rosparam file="$(find aerial_robot_base)/config/d_rook/TeleopNavigationConfig.yaml" command="load" />

    ############### remap topic ######################
    <remap from="/navigator/joy_stick_command" to="/joy" />

    <remap from="/imu/kduino/imu" to="/kduino/imu" />
    <remap from="/config_cmd" to="/kduino/msp_cmd" />
    <remap from="arming_ack" to="/kduino/arming_ack" />
  </node>

</launch>



